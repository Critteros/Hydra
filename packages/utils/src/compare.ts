import { isPrimitive, isNumber, isString, select } from '@/object';
import type { AnyObject, Leaves } from '@/types/objects';

function _shortenBigint(value: bigint): number {
  if (value < 0) {
    return -1;
  }
  if (value > 0) {
    return 1;
  }
  return 0;
}

type _CompareFn<T, R = number> = (first: T, second: T) => R;
function _makeCompareFunction(config: {
  compareNumbers: _CompareFn<number>;
  compareStrings: _CompareFn<string>;
  compareBigInts: _CompareFn<bigint, bigint>;
}) {
  const { compareBigInts, compareStrings, compareNumbers } = config;
  const func = <Item>(objPath?: Item extends AnyObject ? Leaves<Item> : string) => {
    return (first: Item, second: Item): number => {
      if (first == null || typeof first === 'symbol') {
        const objType = typeof first;
        throw new TypeError(
          `Comparison is not supported between objects of ${objType === 'object' ? 'null' : objType}`,
        );
      }

      if (isNumber(first)) {
        return compareNumbers(first, second as number);
      }

      if (isString(first)) {
        return compareStrings(first, second as string);
      }

      if (typeof first === 'bigint') {
        const value = compareBigInts(first, second as bigint);
        return _shortenBigint(value);
      }

      if (typeof first === 'object') {
        if (objPath === undefined) {
          throw new TypeError('Object path has to be provided when working with objects');
        }

        return func()(
          // Type inference would not work here nicely with this amount of generics
          select(first as AnyObject, objPath as never),
          select(second as AnyObject, objPath as never),
        );
      }

      throw new TypeError(`Sorting not supported for ${typeof first}`);
    };
  };

  return func;
}

/**
 * Helper function to sort items in ascending order
 *
 * Works with:
 *
 * - Numbers
 * - Bigints
 * - Strings
 * - Objects (if `objPath` is supped for one of the above types)
 *
 * @example
 *   // returns [1, 2, 3, 5, 7, 8]
 *   [5, 3, 7, 8, 1, 2].toSorted(ascending());
 *
 *   // returns ['a', 'b', 'c', 'd', 'x', 'z']
 *   ['b', 'd', 'z', 'x', 'c', 'a'].toSorted(ascending());
 *
 *   // returns [{ key: 1 }, { key: 2 }]
 *   [{ key: 2 }, { key: 1 }].toSorted(ascending('key'));
 *
 * @param objPath - Object path to retrieve key for compare
 * @throws {TypeError} If used with types that does not support sorting like symbol
 * @throws {TypeError} If used with object without providing the `objPath` parameter
 */
export const ascending = _makeCompareFunction({
  compareNumbers: (first, second) => first - second,
  compareStrings: (first, second) => first.localeCompare(second),
  compareBigInts: (first, second) => first - second,
});

/**
 * Helper function to sort items in descending order
 *
 * Works with:
 *
 * - Numbers
 * - Bigints
 * - Strings
 * - Objects (if `objPath` is supped for one of the above types)
 *
 * @example
 *   // returns [8, 7, 5, 3, 2, 1]
 *   [5, 3, 7, 8, 1, 2].toSorted(ascending());
 *
 *   // returns ['z', 'x', 'd', 'c', 'b', 'a']
 *   ['b', 'd', 'z', 'x', 'c', 'a'].toSorted(ascending());
 *
 *   // returns [{ key: 2 }, { key: 1 }]
 *   [{ key: 1 }, { key: 2 }].toSorted(ascending('key'));
 *
 * @param objPath - Object path to retrieve key for compare
 * @throws {TypeError} If used with types that does not support sorting like symbol
 * @throws {TypeError} If used with object without providing the `objPath` parameter
 */
export const descending = _makeCompareFunction({
  compareNumbers: (first, second) => second - first,
  compareStrings: (first, second) => second.localeCompare(first),
  compareBigInts: (first, second) => second - first,
});
