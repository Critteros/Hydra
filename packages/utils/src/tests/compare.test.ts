import { describe, expect, test } from 'vitest';

import { ascending, descending } from '../compare';

describe.concurrent('Compare utilities', () => {
  describe.concurrent('sorting helper', () => {
    test('works with basic number array', () => {
      const arr = [5, 3, 7, 8, 1, 2];
      expect(arr.toSorted(ascending())).toStrictEqual([1, 2, 3, 5, 7, 8]);
      expect(arr.toSorted(descending())).toStrictEqual([8, 7, 5, 3, 2, 1]);
    });

    test('works with array of strings', () => {
      const arr = ['b', 'd', 'z', 'x', 'c', 'a'];
      expect(arr.toSorted(ascending())).toStrictEqual(['a', 'b', 'c', 'd', 'x', 'z']);
      expect(arr.toSorted(descending())).toStrictEqual(['z', 'x', 'd', 'c', 'b', 'a']);
    });

    test('works with the array of bigints', () => {
      const arr = [BigInt('10'), BigInt('7'), BigInt('1'), BigInt('2'), BigInt('100')];
      expect(arr.toSorted(ascending())).toStrictEqual([
        BigInt('1'),
        BigInt('2'),
        BigInt('7'),
        BigInt('10'),
        BigInt('100'),
      ]);
      expect(arr.toSorted(descending())).toStrictEqual([
        BigInt('100'),
        BigInt('10'),
        BigInt('7'),
        BigInt('2'),
        BigInt('1'),
      ]);
    });

    test('works with array of objects', () => {
      const arr = [
        {
          val: 3,
          x: 'v',
        },
        {
          val: 7,
          x: 'x',
        },
        {
          val: 1,
          x: '0',
        },
      ];
      expect(arr.toSorted(ascending('val'))).toStrictEqual([
        {
          val: 1,
          x: '0',
        },
        {
          val: 3,
          x: 'v',
        },
        {
          val: 7,
          x: 'x',
        },
      ]);
      expect(arr.toSorted(descending('val'))).toStrictEqual([
        {
          val: 7,
          x: 'x',
        },
        {
          val: 3,
          x: 'v',
        },
        {
          val: 1,
          x: '0',
        },
      ]);
    });

    test('key path has to be provided when sorting objects', () => {
      const arr = [
        {
          x: 3,
        },
        {
          x: 1,
        },
        {
          x: 100,
        },
      ];

      expect(() => arr.toSorted(ascending())).toThrowErrorMatchingInlineSnapshot(
        `[TypeError: Object path has to be provided when working with objects]`,
      );
      expect(() => arr.toSorted(descending())).toThrowErrorMatchingInlineSnapshot(
        `[TypeError: Object path has to be provided when working with objects]`,
      );
    });

    test.concurrent.for`
      testValue
      ${Symbol('test')}
      ${null}
      ${true}
      ${false}
    `('is not supported for $testValue', ({ testValue }, { expect }) => {
      const arr = [testValue, testValue] as unknown[];
      expect(() => arr.toSorted(ascending())).toThrowErrorMatchingSnapshot();
      expect(() => arr.toSorted(descending())).toThrowErrorMatchingSnapshot();
    });

    test('is stable sort', () => {
      const arr = [
        {
          x: BigInt('10'),
          val: 10,
        },
        {
          x: BigInt('10'),
          val: 1,
        },
        {
          x: BigInt('10'),
          val: 7,
        },
      ];
      expect(arr.toSorted(ascending('x'))).toStrictEqual([
        {
          x: BigInt('10'),
          val: 10,
        },
        {
          x: BigInt('10'),
          val: 1,
        },
        {
          x: BigInt('10'),
          val: 7,
        },
      ]);
      expect(arr.toSorted(descending('x'))).toStrictEqual([
        {
          x: BigInt('10'),
          val: 10,
        },
        {
          x: BigInt('10'),
          val: 1,
        },
        {
          x: BigInt('10'),
          val: 7,
        },
      ]);
    });
  });
});
