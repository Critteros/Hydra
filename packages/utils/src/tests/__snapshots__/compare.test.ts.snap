// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Compare utilities > sorting helper > is not supported for Symbol(test) 1`] = `[TypeError: Comparison is not supported between objects of symbol]`;

exports[`Compare utilities > sorting helper > is not supported for Symbol(test) 2`] = `[TypeError: Comparison is not supported between objects of symbol]`;

exports[`Compare utilities > sorting helper > is not supported for false 1`] = `[TypeError: Sorting not supported for boolean]`;

exports[`Compare utilities > sorting helper > is not supported for false 2`] = `[TypeError: Sorting not supported for boolean]`;

exports[`Compare utilities > sorting helper > is not supported for null 1`] = `[TypeError: Comparison is not supported between objects of null]`;

exports[`Compare utilities > sorting helper > is not supported for null 2`] = `[TypeError: Comparison is not supported between objects of null]`;

exports[`Compare utilities > sorting helper > is not supported for true 1`] = `[TypeError: Sorting not supported for boolean]`;

exports[`Compare utilities > sorting helper > is not supported for true 2`] = `[TypeError: Sorting not supported for boolean]`;
